equinox/nn/_attention.py

```diff
@@ -325,6 +377,42 @@ def __call__(
                 raise ValueError(
                     "process_heads must not change the shape of the heads."
                 )
+        if state is None:
+            state_length = None
+            causal_mask_offset = 0
+        else:
+            if self.kv_cache is None:
+                raise ValueError(
+                    "State was provided, but cannot use autoregressive decoding "
+                    "without specifying "
+                    "`MultiheadAttention(..., kv_cache=...)`. "
+                    "See `equinox.nn.StandardKVCache` for an example."
+                )
+
+            key_state, value_state, index, state = self.kv_cache(
+                key_heads, value_heads, state
+            )
```

Comment:
I've just realised -- is this actually doing the correct thing if we also have a RoPE? I don't think it is.

Hmm, do we need to also pass a start index to the RoPE to make this work?

Reply:
I don't quite follow. 

RoPE is applied _after_ the linear projections:

```python
        query_heads = self._project(self.query_proj, query)
        key_heads = self._project(self.key_proj, key_)
        value_heads = self._project(self.value_proj, value)

        if process_heads is not None:
            q_shape, k_shape, v_shape = (
                query_heads.shape,
                key_heads.shape,
                value_heads.shape,
            )
            query_heads, key_heads, value_heads = process_heads(
                query_heads, key_heads, value_heads
            )
```

And after RoPE is applied, the keys and values are put into the KV cache. This means that the "RoPE-applied" keys and values are stored/updated. 

Or, did you mean that we should:

1. Store/update the KV cache
2. Apply `process_heads` *after* fetching the KV from the cache:

```python
            if self.kv_cache is None:
                raise ValueError(
                    "State was provided, but cannot use autoregressive decoding "
                    "without specifying "
                    "`MultiheadAttention(..., kv_cache=...)`. "
                    "See `equinox.nn.StandardKVCache` for an example."
                )

            key_state, value_state, index, state = self.kv_cache(
                key_heads, value_heads, state
            )
            _check_kv_shapes(key_state, value_state, key_heads, value_heads)
            state_length, _, _ = key_state.shape

            causal_mask_offset = index
            key_heads = key_state
            value_heads = value_state
            kv_seq_length = state_length

            # PROCESS HEADS HERE INSTEAD
```

Reply:
I've checked out the code from LLaMA3 as a reference, and they apply RoPE before inserting the arrays into the KV cache.

Reply:
Right, but RoPe uses a bunch of precomputed sines and cosines whose values oscillate according to how far along the sequence they are: https://github.com/patrick-kidger/equinox/blob/d3a01c8d1b5531f193f11514c0189104b8f62136/equinox/nn/_embedding.py#L186-L189

However when using KV caching, then we won't be adjusting that `arange` appropriately to account for how far along the sequence we are.